### Sección 1: Conceptos Teóricos (40 puntos)

**(5 pts) ¿Cuál es la diferencia entre una regresión lineal simple y una regresión lineal múltiple? Da un ejemplo de cada una.**

La regresión lineal simple utiliza una variable independiente para predecir la dependiente. Un ejemplo sería utilizar la estatura para predecir el peso.

Modelo: $$ \hat y = \beta_0 + \beta_1 x_1 $$

La regresión lineal múltiple utiliza varias variables independientes para predecir la dependiente. Un ejemplo sería usar la temperatura, sensación térmica, humedad, día de la semana y estación del año para pronosticar la cantidad de bicis que se rentan en una ciudad.

Modelo: $$ \hat y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \ ... \ + \beta_n x_n $$

**(5 pts) Explica el problema de overfitting y cómo podrías mitigarlo**

Ocurre cuando el modelo obtenido memoriza los datos en lugar de aprender, por lo que no puede generalizar pues es muy malo en datos no vistos por el modelo. Para mitigarlo se puede hacer cross validation, donde una parte de los datos son para entrenar al modelo y otra para probarlo (datos no vistos) e intentas que mejore el modelo en el test. Además utilizar regularización Ridge o Lasso en una regresión lineal puede ayudar a no hacer overfitting.

**(5 pts) En regresión polinomial, ¿por qué es necesario realizar un análisis de validación cruzada antes de elegir el grado del polinomio?**

La razón prinicpal es para evitar el problema de overfitting, pues no te interesa que el modelo memorice, sino que sea capaza de predecir con base en nuevas entradas. Esto se debe a que un modelo de grado muy alto puede ajustarse muy bien a los datos, sin embargo, ser incapaz de generalizar y hacer buenas predicciones

**(5 pts) ¿Qué es una prueba de hipótesis en el contexto de regresión lineal? Explica cómo se interpreta el p-value de un coeficiente.**

La prueba de hipótesis nos dice si las $\beta$'s son significativas o no, para ello calcula el estadístico de t de estas el cual nos dice a cuantas desviaciones estándar esta $\beta$ de 0, y se obtiene el p-value asociado al valor de t obtenido, y a partir de esto decide si la variable es significativa o no.

Si la variable tiene un p-value mayor a 0.05 (o valor de umbral establecido), significa que la variable no es significativa para la regresión, lo cual indica que muy posiblemente la variable no aporte información al modelo. Si es menor a 0.05 el p-value de la variable significa que si aporta al modelo y es una variable útil.

**(5 pts) ¿Por qué convertir variables categóricas en dummies puede mejorar el desempeño de un modelo? ¿En qué casos podríamos usar una codificación diferente?**

Puede ayudar a darle más información al modelo que sea útil para predecir. Por ejemplo, si tienes una columna que sea el día de la semana, lo necesitas de forma numércia para que sea utilizable para el modelo, entonces tendrías 7 columnas, donde es 1 si ese día de la semana y 0 si no, esto ayuda al modelo a captar si afecta que sea algún día en particular lo cual puede ser bueno.

Otra forma de hacerlo es transformar variables numéricas a categóricas a través de los bins, esto permite dividir tus variables numéricas en cuantiles hacinedola categórica. Tener estas categorías pueden ayudar a predecir al modelo basandose en si una observación pertence a algun cuantil. Por ejemplo si se tiene ingreso como variable, esta estrategia clasificaría a la gente con ingreso muy alto en un grupo, lo cual ayuda pues seguramente el ingreso muy alto tiene un comportamiento distinto en comparación con alguien de ingreso muy bajo.

**(5 pts) Explica como se obtienen los coeficientes de una regresión lineal con decenso en gradiente (tanto para lineal como polinomio), me tiene que quedar claro que le entiendes**

El descenso en gradiente es un proceso iterativo el cual busca minimizar tu error (función de pérdida) actualizando los valores de las $\beta$'s de tu regresión a través de derivadas parciales.

Tienes la función de pérdida: $L = \frac{1}{2} \sum (\hat y - y)^2$

Para este ejemplo $\hat y = \beta_0 + \beta_1 x_1$

Tu buscas:
$$
min \quad \frac{1}{2} \sum (\hat y - y)^2
$$

Para eso realizas de forma iterativa lo siguiente ($\alpha$ es tasa de aprendizaje): $\beta_1 = \beta - \alpha \frac{\partial L}{\partial \beta}$

Cada $\beta$ se actualiza restándole la derivada parcial de $L$ multiplicada por la tasa de aprendizaje $\alpha$, la cual ayuda a que el cambio en la $\beta$ no sea tan grande. Esta nueva $\beta$ se utiliza para recalcular la función de pérdida y se vuelve a actualizar en cada iteración, restando nuevamente la derivada parcial multiplicada por la tasa de aprendizaje. Este proceso se repite hasta que el modelo converge. La derivada parcial se hace sobre todas las $\beta$'s y todas se actualizan.

El método funciona porque al hacer la resta, el signo negativo invierte la dirección de la pendiente obtenida en la derivada después de cada iteración, ayudando a la convergencia.

**(5 pts) Explica que es el teorema de Frisch-Waugh-Lovell**

Para una regresión lineal multiple $ \hat y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \beta_3 x_3 $, la $\beta_1$ puede obtenerse a través de un método, logrando explicar el efecto causal de una variable.

$$
\hat y = \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \beta_3 x_3
$$
Esta es la regresión lineal original con todas las variables, y se busca encontrar el efecto causal de la variable $x_1$.

$$
\hat y = \theta_0 + \theta_2 x_2 + \theta_3 x_3
$$
Para ello haces la regresión lineal para $y$ sin utilizar $x_1$.

$$
y - \hat y = \text{residuales y}
$$
Y obtienes los residuales entre tu regresión sin $x_1$ y los valores reales de $y$. Estos residuales de $y$ es lo que no pueden explicar $x_2$ y $x_3$ de $y$.


$$
\hat {x_1} = \gamma_0 + \gamma_2 x_2 + \gamma_3 x_3
$$
Posteriormente realizas una regresión lineal para $x_1$ con las variables $x_2$ y $x_3$.

$$
x_1 - \hat {x_1} = \text{residuales x}
$$
Y calculas los residuales entre tu predicción de $\hat {x_1}$ y $x_1$ real. Estos residuales de $x_1$ es lo que no pueden explicar  $x_2$ y $x_3$ de $x_1$

$$
y - \hat y = \beta_1(x_1 - \hat {x_1})
$$
Por útlimo realizas un regresión de los residuales de $y$ contra los residuales de $x_1$, y la $B_1$ resultante es la misma que la $B_1$ de la regresión original, lo cual sería el efecto causal de $x_1$ sobre $y$.


**(5 pts) Explica que es K-nearest neighboors**

Es un modelo que puede utilizarse para predecir o clasificar, basadóse en el número de vecinos más cercanos según las características del punto.

El modelo funciona eligiendo un número de vecinos (K), para un nuevo dato calcula las distancias con todos los datos que son parte de tu entrenamiento, y selecciona a los K vecinos más cercanos. Si elegista 3 vecinos, entonces selecciona a los 3 puntos más cercanos. Y la predicción es el promedio de los valores de los K vecinos más cercanos.

Si el vecino 1 = 8, vecino 2 = 15 y vecino 3 = 4, entonces la predicción sería el promedio que es 9.

### Sección 2 (20 puntos)

(10 pts) Dado el siguiente dataset de casas:

```
data = pd.DataFrame({
    "Tamaño_m2": [50, 80, 120, 200, 150, 90, 175, 60, 220, 130],
    "Habitaciones": [1, 2, 3, 4, 3, 2, 3, 1, 5, 3],
    "Precio_1000s": [110, 170, 250, 400, 270, 200, 330, 190, 600, 380]  })
````



a) Ajusta una regresión lineal para predecir el precio en función del tamaño y número de habitaciones, dime el R2 y coeficientes.


b) Imagina que el coeficiente de Tamaño es 2.5 y el coeficiente de Habitaciones es 20, ¿cómo interpretarías estos valores?  

(20 pts) Imagina que entrenas un modelo de regresión polinomial de grado 5 y obtienes un 𝑅2 de 0.98 en el conjunto de entrenamiento y 0.65 en el conjunto de prueba.

- a) ¿Qué problema podrías estar enfrentando?
- b) ¿Cómo lo solucionarías sin reducir demasiado la capacidad del modelo?


In [1]:
import numpy as np
import pandas as pd
from sklearn.metrics import r2_score
import statsmodels.formula.api as smf


In [3]:
data = pd.DataFrame({
    "Tamaño_m2": [50, 80, 120, 200, 150, 90, 175, 60, 220, 130],
    "Habitaciones": [1, 2, 3, 4, 3, 2, 3, 1, 5, 3],
    "Precio_1000s": [110, 170, 250, 400, 270, 200, 330, 190, 600, 380]})
data.head()

Tamaño_m2  Habitaciones  Precio_1000s
0         50             1           110
1         80             2           170
2        120             3           250
3        200             4           400
4        150             3           270

In [4]:
model = smf.ols('Precio_1000s ~ Tamaño_m2 + Habitaciones', data=data).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:           Precio_1000s   R-squared:                       0.866
Model:                            OLS   Adj. R-squared:                  0.828
Method:                 Least Squares   F-statistic:                     22.66
Date:                Tue, 18 Feb 2025   Prob (F-statistic):           0.000876
Time:                        20:49:36   Log-Likelihood:                -53.255
No. Observations:                  10   AIC:                             112.5
Df Residuals:                       7   BIC:                             113.4
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
Intercept       -0.0869     47.302     -0.002   

/opt/anaconda3/lib/python3.12/site-packages/scipy/stats/_stats_py.py:1971: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=10
  k, _ = kurtosistest(a, axis)


In [5]:
r2_score(data['Precio_1000s'], model.predict(data[['Tamaño_m2', 'Habitaciones']]))

0.8661894650001757

## Inciso a)
El r^2 del modelo es de 0.866, lo que significa que explica un 86.6% de la variacón de los datos.

Coeficientes:

- Intercepto: -0.0869
- Tamaño_m2 = 0.4469
- Habitaciones = 86.339

## Inciso b

b) Imagina que el coeficiente de Tamaño es 2.5 y el coeficiente de Habitaciones es 20, ¿cómo interpretarías estos valores?  

Por cada metro cuadrado extra, el precio de la casa aumenta 2,500, y por cada habitación extra el valor de la casa aumenta 20,000.

(20 pts) Imagina que entrenas un modelo de regresión polinomial de grado 5 y obtienes un 𝑅2 de 0.98 en el conjunto de entrenamiento y 0.65 en el conjunto de prueba.

- a) ¿Qué problema podrías estar enfrentando?
- b) ¿Cómo lo solucionarías sin reducir demasiado la capacidad del modelo?

En este caso sería un problema de overfitting pues funciona muy bien en los datos de entrenamiento pero en los de prueba realiza malas predicciones.

Para solucionar el problema se pueden realizar varias cosas, una sería usar regularización Ridge o Lasso, otra hacer un análisis bivariado de las variables contra el target para ver su relación y saber si vale la pena hacer transformaciones, o analizar la correlación entre variables y ver si hay algunas que no aporten al modelo y quitarlas.

Ya por último se puede reducir el grado del polinomio pero esto si podría afcetar su capacidad.

### Sección 3: Implementación de Código (40 puntos)


1- En el dataset Advertising.csv estan las ventas de una empresa y hay 3 canales de distribuición:
    
- radio
- Televisión
- Periodico


La empresa quiere simplificar su operción y eliminar aquellos canales que no traen ventas.

Dime que canal o canales no sirven con una prueba de hipotesis y una regresión


    

In [9]:
df = pd.read_csv('Advertising.csv')
df.head()

Unnamed: 0     TV  radio  newspaper  sales
0           1  230.1   37.8       69.2   22.1
1           2   44.5   39.3       45.1   10.4
2           3   17.2   45.9       69.3    9.3
3           4  151.5   41.3       58.5   18.5
4           5  180.8   10.8       58.4   12.9

In [11]:
model = smf.ols('sales ~ TV + radio + newspaper', data=df).fit()
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                  sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Tue, 18 Feb 2025   Prob (F-statistic):           1.58e-96
Time:                        21:05:31   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      2.9389      0.312      9.422      0.0

Con esta regresión lineal vemos como cada una de las variables afecta a sales a través de los coeficientes. Tanto TV como radio son postivas, por lo que a mayor cantidad en esas variables las ventas aumentan. Podemos notar que newspaper tiene un coeficiente negativo, lo cual significa que a mayor en newspaper nuestras ventas se reducen, lo que significa que no deberíamos usar newspaper pues afecta a nuestras ventas.

Además analizando los p-values de nuestras variables vemos que el de newspaper es 0.860, lo que implica que no es significativo, por lo que no aporta información útil para las predicciones del modelo, por lo que se puede quitar.

En conclusión la empresa debería quitar el canal de vetas de newspapaer pues afceta de forma negativa las ventas y su p-value es muy alto por lo que no es significativa para nuestro modelo.